src/vis.3
.\"	$OpenBSD: vis.3,v 1.23 2005/08/28 19:51:27 millert Exp $
.\"
.\" Copyright (c) 1989, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd $Mdocdate: May 31 2007 $
.Dt VIS 3
.Os
.Sh NAME
.Nm vis ,
.Nm strvis ,
.Nm strnvis ,
.Nm strvisx
.Nd visually encode characters
.Sh LIBRARY
.ds str-Lb-libbsd Utility functions from BSD systems (libbsd, \-lbsd)
.Lb libbsd
.Sh SYNOPSIS
.In stdlib.h
.In bsd/vis.h
.Ft char *
.Fn vis "char *dst" "int c" "int flag" "int nextc"
.Ft int
.Fn strvis "char *dst" "const char *src" "int flag"
.Ft int
.Fn strnvis "char *dst" "const char *src" "size_t size" "int flag"
.Ft int
.Fn strvisx "char *dst" "const char *src" "size_t len" "int flag"
.Sh DESCRIPTION
The
.Fn vis
function copies into
.Fa dst
a string which represents the character
.Fa c .
If
.Fa c
needs no encoding, it is copied in unaltered.
The string is NUL terminated and a pointer to the end of the string is
returned.
The maximum length of any encoding is four
characters (not including the trailing NUL);
thus, when
encoding a set of characters into a buffer, the size of the buffer should
be four times the number of characters encoded, plus one for the trailing
NUL.
The
.Fa flag
parameter is used for altering the default range of
characters considered for encoding and for altering the visual
representation.
The additional character,
.Fa nextc ,
is only used when selecting the
.Dv VIS_CSTYLE
encoding format (explained below).
.Pp
The
.Fn strvis ,
.Fn strnvis
and
.Fn strvisx
functions copy into
.Fa dst
a visual representation of
the string
.Fa src .
The
.Fn strvis
function encodes characters from
.Fa src
up to the first NUL.
The
.Fn strnvis
function encodes characters from
.Fa src
up to the first NUL or the end of
.Fa dst ,
as indicated by
.Fa size .
The
.Fn strvisx
function encodes exactly
.Fa len
characters from
.Fa src
(this
is useful for encoding a block of data that may contain NULs).
All three forms NUL terminate
.Fa dst ,
except for
.Fn strnvis
when
.Fa size
is zero, in which case
.Fa dst
is not touched.
For
.Fn strvis
and
.Fn strvisx ,
the size of
.Fa dst
must be four times the number
of characters encoded from
.Fa src
(plus one for the NUL).
.Fn strvis
and
.Fn strvisx
return the number of characters in
.Fa dst
(not including the trailing NUL).
.Fn strnvis
returns the length that
.Fa dst
would become if it were of unlimited size (similar to
.Xr snprintf 3
or
.Xr strlcpy 3 ) .
This can be used to detect truncation but it also means that
the return value of
.Fn strnvis
must not be used without checking it against
.Fa size .
.Pp
The encoding is a unique, invertible representation composed entirely of
graphic characters; it can be decoded back into the original form using
the
.Xr unvis 3
or
.Xr strunvis 3
functions.
.Pp
There are two parameters that can be controlled: the range of
characters that are encoded, and the type
of representation used.
By default, all non-graphic characters
except space, tab, and newline are encoded
(see
.Xr isgraph 3 ) .
The following flags
alter this:
.Bl -tag -width VIS_WHITEX
.It Dv VIS_GLOB
Also encode magic characters recognized by
.Xr glob 3
.Pf ( Ql * ,
.Ql \&? ,
.Ql \&[ )
and
.Ql # .
.It Dv VIS_SP
Also encode space.
.It Dv VIS_TAB
Also encode tab.
.It Dv VIS_NL
Also encode newline.
.It Dv VIS_WHITE
Synonym for
.Dv VIS_SP
\&|
.Dv VIS_TAB
\&|
.Dv VIS_NL .
.It Dv VIS_SAFE
Only encode
.Dq unsafe
characters.
These are control characters which may cause common terminals to perform
unexpected functions.
Currently this form allows space,
tab, newline, backspace, bell, and return -- in addition
to all graphic characters -- unencoded.
.El
.Pp
There are three forms of encoding.
All forms use the backslash
.Ql \e
character to introduce a special
sequence; two backslashes are used to represent a real backslash.
These are the visual formats:
.Bl -tag -width VIS_CSTYLE
.It (default)
Use an
.Ql M
to represent meta characters (characters with the 8th
bit set), and use a caret
.Ql ^
to represent control characters (see
.Xr iscntrl 3 ) .
The following formats are used:
.Bl -tag -width xxxxx
.It Dv \e^C
Represents the control character
.Ql C .
Spans characters
.Ql \e000
through
.Ql \e037 ,
and
.Ql \e177
(as
.Ql \e^? ) .
.It Dv \eM-C
Represents character
.Ql C
with the 8th bit set.
Spans characters
.Ql \e241
through
.Ql \e376 .
.It Dv \eM^C
Represents control character
.Ql C
with the 8th bit set.
Spans characters
.Ql \e200
through
.Ql \e237 ,
and
.Ql \e377
(as
.Ql \eM^? ) .
.It Dv \e040
Represents
.Tn ASCII
space.
.It Dv \e240
Represents Meta-space.
.El
.Pp
.It Dv VIS_CSTYLE
Use C-style backslash sequences to represent standard non-printable
characters.
The following sequences are used to represent the indicated characters:
.Bd -unfilled -offset indent
.Li \ea Tn  - BEL No (007)
.Li \eb Tn  - BS No (010)
.Li \ef Tn  - NP No (014)
.Li \en Tn  - NL No (012)
.Li \er Tn  - CR No (015)
.Li \es Tn  - SP No (040)
.Li \et Tn  - HT No (011)
.Li \ev Tn  - VT No (013)
.Li \e0 Tn  - NUL No (000)
.Ed
.Pp
When using this format, the
.Fa nextc
parameter is looked at to determine
if a NUL character can be encoded as
.Ql \e0
instead of
.Ql \e000 .
If
.Fa nextc
is an octal digit, the latter representation is used to
avoid ambiguity.
.It Dv VIS_OCTAL
Use a three digit octal sequence.
The form is
.Ql \eddd
where
.Ar d
represents an octal digit.
.El
.Pp
There is one additional flag,
.Dv VIS_NOSLASH ,
which inhibits the
doubling of backslashes and the backslash before the default
format (that is, control characters are represented by
.Ql ^C
and
meta characters as
.Ql M-C ) .
With this flag set, the encoding is
ambiguous and non-invertible.
.Sh SEE ALSO
.Xr unvis 1 ,
.Xr vis 1 ,
.Xr snprintf 3 ,
.Xr strlcpy 3 ,
.Xr unvis 3
.Sh HISTORY
The
.Fn vis ,
.Fn strvis
and
.Fn strvisx
functions first appeared in
.Bx 4.4 .
The
.Fn strnvis
function first appeared in
.Ox 2.9 .


ChangeLog
2008-03-16  Guillem Jover  <guillem@debian.org>

	* include/vis.h: Remove UC Berkeley advertising clause, as per
	<ftp://ftp.cs.berkeley.edu/pub/4bsd/README.Impt.License.Change>.
	* src/vis.c: Likewise.
	* src/unvis.c: Likewise.
	* src/heapsort.c: Likewise.
	* src/linkaddr.c: Likewise.
	* man/fgetln.3: Likewise.

2007-12-18  Petr Salinger  <Petr.Salinger@seznam.cz>

	* include/bsd/queue.h: sync with current glibc variant on GNU/kFreeBSD

2006-11-08  Aurelien Jarno  <aurelien@aurel32.net>

	* src/link_addr.c: New file.
	* include/bsd/if_dl.h: Add link_addr.c declarations.
	* Makefile (LIB_SRCS): Add link_addr.c.
	* Versions: add link_addr and link_ntoa

2006-08-25  Petr Salinger  <Petr.Salinger@seznam.cz>

	* src/humanize_number.c: New file.
	* include/libutil.h: New file with humanize_number declarations.
	* Makefile (LIB_SRCS): Add humanize_number.c.
	* Versions: Add humanize_number.

2006-03-29  Robert Millan  <rmh@aybabtu.com>

	* src/heapsort.c: New file.
	* include/bsd/stdlib.h: Add heapsort declaration.
	* Makefile (LIB_SRCS): Add heapsort.c.
	* Versions: Add heapsort.

2006-03-19  Robert Millan  <rmh@aybabtu.com>

	* src/vis.c: New file.
	* src/unvis.c: Ditto.
	* include/vis.h: Add vis/unvis declarations.
	* Makefile (LIB_SRCS): Add vis.c/unvis.c.
	* Versions: Add vis/unvis functions.

2006-03-17  Robert Millan  <rmh@aybabtu.com>

	* src/progname.c: New file.
	* include/bsd/stdlib.h: Add {get,set}progname declarations.
	* Makefile (LIB_SRCS): Add 'progname.c'.
	* Versions: Add getprogname and setprogname.

2006-02-15  Guillem Jover  <guillem@debian.org>

	* include/bsd/cdefs.h: Disable __unused, it conflicts with a struct
	member in a Linux header.

2006-02-14  Guillem Jover  <guillem@debian.org>

	* include/bsd/stdlib.h: Remove <sys/types.h> and <stdio.h>.
	* include/bsd/getopt.h: Protect header from multiple inclusion.

2006-02-14  Guillem Jover  <guillem@debian.org>

	* include/bsd/bsd.h: Remove commented out macros.

2006-02-14  Guillem Jover  <guillem@debian.org>

	* include/bsd/bsd.h (setproctitle, __dead2, __unused, __printflike)
	(__FBSDID): Moved to ...
	* include/bsd/cdefs.h: ... here.
	(__unused, __printflike) [__GNUC__]: Use proper __attribute__.
	* Makefile (LIB_INCLUDES): Add 'cdefs.h'.

2006-02-14  Guillem Jover  <guillem@debian.org>

	* arc4random.c: Move to ...
	* src/arc4random.c: ... here.
	* bsd_getopt.c: Move to ...
	* src/bsd_getopt.c:: ... here.
	* err.c: Move to ...
	* src/err.c:: ... here.
	* fgetln.c: Move to ...
	* src/fgetln.c:: ... here.
	* fmtcheck.c: Move to ...
	* src/fmtcheck.c:: ... here.
	* inet_net_pton.c: Move to ...
	* src/inet_net_pton.c:: ... here.
	* strlcat.c: Move to ...
	* src/strlcat.c:: ... here.
	* strlcpy.c: Move to ...
	* src/strlcpy.c:: ... here.
	* md5c.c: Move to ...
	* src/md5c.c:: ... here.
	* md5.copyright: Move to ...
	* src/md5.copyright:: ... here.
	* Makefile (LIB_SRCS): Automatically prefix the src/ dir.

2006-02-14  Guillem Jover  <guillem@debian.org>

	* Makefile (LIB_SRCS): Line wrap.
	(LIB_INCLUDES): List only the files, automatically prefix the dir.
	(LIB_MANS): Likewise.

2006-02-13  Robert Millan  <rmh@aybabtu.com>

	* bsd_getopt.c: New file.
	* include/bsd/getopt.h: Likewise.
	* Makefile (LIB_SRCS): Add 'bsd_getopt.c'.
	(LIB_INCLUDES): Add 'include/bsd/getopt.h'.
	* Versions: Added bsd_getopt and optreset.

2006-02-10  Robert Millan  <rmh@aybabtu.com>

	* err.c: New file.
	* include/bsd/err.h: Likewise.
	* Makefile (LIB_SRCS): Add 'err.c'.
	(LIB_INCLUDES): Add 'include/bsd/err.h'.
	* Versions: Added errc, warnc, verrc and vwarn.

2005-12-19  Aurelien Jarno <aurel32@debian.org>

	* man/arc4random.3: New file.
	* man/fgetln.3: Likewise.
	* man/fmtcheck.3: Likewise.
	* man/strlcpy.3: Likewise.
	* fmtcheck.c:: Likewise.
	* include/stdlib.h: Added fmtcheck declaration.
	* Versions: Added fmtcheck.

2005-11-01  Guillem Jover  <guillem@debian.org>

	* ip_icmp.h: Move to ...
	* include/bsd/ip_icmp.h: ... here.
	* md5.h: Move to ...
	* include/bsd/md5.h: ... here.
	* bsd.h: Move to ...
	* include/bsd/bsd.h: ... here. Split contents ...
	* include/bsd/random.h: ... here ...
	* include/bsd/queue.h: ... here ...
	* include/bsd/string.h: ... and here.
	* md5c.c: Change md5.h include line.
	* Makefile (MK_CFLAGS): Include from the proper dir.

2005-08-03  Hector Garcia  <hector@debian.org>

	* Makefile: Fixed to place soft links instead in soname, and
	shared_so instead of copying files.
	* md5.h (MD5End, MD5File, MD5FileChunk, MD5Data): Commented out
	functions not available on 'md5c.c'.

2005-07-26  Guillem Jover  <guillem@debian.org>

	* Makefile (LIB_INCLUDES): Add 'md5.h'.

2005-07-26  Guillem Jover  <guillem@debian.org>

	* Versions (LIBBSD_0.0): Add MD5Init, MD5Update, MD5Pad and
	MD5Final.

2005-07-26  Hector Garcia  <hector@debian.org>

	* Makefile (LIB_SRCS): Add 'md5c.c'.
	* md5c.c: Changed include to 'md5.h'.
	* md5.h: Replaced with the system header 'sys/md5.h'.

2005-07-25  Guillem Jover  <guillem@debian.org>

	* fgetln.c (fgetln): Reindent, remove commented code and translate
	variable names to english.

2005-07-25  Guillem Jover  <guillem@debian.org>

	* Versions: New file.
	* Makefile: Add versioned symbols support.



src/vis.c
/*	$OpenBSD: vis.c,v 1.18 2005/08/29 18:38:41 otto Exp $ */
/*-
 * Copyright (c) 1989, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/types.h>
#include <limits.h>
#include <ctype.h>
#include <string.h>
#include <stdio.h>
#include <vis.h>

#define	isoctal(c)	(((u_char)(c)) >= '0' && ((u_char)(c)) <= '7')
#define	isvisible(c)							\
	(((u_int)(c) <= UCHAR_MAX && isascii((u_char)(c)) &&		\
	(((c) != '*' && (c) != '?' && (c) != '[' && (c) != '#') ||	\
		(flag & VIS_GLOB) == 0) && isgraph((u_char)(c))) ||	\
	((flag & VIS_SP) == 0 && (c) == ' ') ||				\
	((flag & VIS_TAB) == 0 && (c) == '\t') ||			\
	((flag & VIS_NL) == 0 && (c) == '\n') ||			\
	((flag & VIS_SAFE) && ((c) == '\b' ||				\
		(c) == '\007' || (c) == '\r' ||				\
		isgraph((u_char)(c)))))

/*
 * vis - visually encode characters
 */
char *
vis(char *dst, int c, int flag, int nextc)
{
	c = (unsigned char)c;

	if (flag & VIS_HTTPSTYLE) {
		/* Described in RFC 1808 */
		if (!(isalnum(c) /* alpha-numeric */
		    /* safe */
		    || c == '$' || c == '-' || c == '_' || c == '.' || c == '+'
		    /* extra */
		    || c == '!' || c == '*' || c == '\'' || c == '('
		    || c == ')' || c == ',')) {
			*dst++ = '%';
			snprintf(dst, 4, (c < 16 ? "0%X" : "%X"), c);
			dst += 2;
			goto done;
		}
	}

	if ((flag & VIS_GLOB) &&
	    (c == '*' || c == '?' || c == '[' || c == '#'))
		;
	else if (isgraph(c) ||
	   ((flag & VIS_SP) == 0 && c == ' ') ||
	   ((flag & VIS_TAB) == 0 && c == '\t') ||
	   ((flag & VIS_NL) == 0 && c == '\n') ||
	   ((flag & VIS_SAFE) && (c == '\b' || c == '\007' || c == '\r'))) {
		*dst++ = c;
		if (c == '\\' && (flag & VIS_NOSLASH) == 0)
			*dst++ = '\\';
		*dst = '\0';
		return (dst);
	}

	if (flag & VIS_CSTYLE) {
		switch(c) {
		case '\n':
			*dst++ = '\\';
			*dst++ = 'n';
			goto done;
		case '\r':
			*dst++ = '\\';
			*dst++ = 'r';
			goto done;
		case '\b':
			*dst++ = '\\';
			*dst++ = 'b';
			goto done;
		case '\a':
			*dst++ = '\\';
			*dst++ = 'a';
			goto done;
		case '\v':
			*dst++ = '\\';
			*dst++ = 'v';
			goto done;
		case '\t':
			*dst++ = '\\';
			*dst++ = 't';
			goto done;
		case '\f':
			*dst++ = '\\';
			*dst++ = 'f';
			goto done;
		case ' ':
			*dst++ = '\\';
			*dst++ = 's';
			goto done;
		case '\0':
			*dst++ = '\\';
			*dst++ = '0';
			if (isoctal(nextc)) {
				*dst++ = '0';
				*dst++ = '0';
			}
			goto done;
		}
	}
	if (((c & 0177) == ' ') || isgraph(c) || (flag & VIS_OCTAL)) {
		*dst++ = '\\';
		*dst++ = ((u_char)c >> 6 & 07) + '0';
		*dst++ = ((u_char)c >> 3 & 07) + '0';
		*dst++ = ((u_char)c & 07) + '0';
		goto done;
	}
	if ((flag & VIS_NOSLASH) == 0)
		*dst++ = '\\';
	if (c & 0200) {
		c &= 0177;
		*dst++ = 'M';
	}
	if (iscntrl(c)) {
		*dst++ = '^';
		if (c == 0177)
			*dst++ = '?';
		else
			*dst++ = c + '@';
	} else {
		*dst++ = '-';
		*dst++ = c;
	}
done:
	*dst = '\0';
	return (dst);
}

/*
 * strvis, strnvis, strvisx - visually encode characters from src into dst
 *
 *	Dst must be 4 times the size of src to account for possible
 *	expansion.  The length of dst, not including the trailing NUL,
 *	is returned.
 *
 *	Strnvis will write no more than siz-1 bytes (and will NULL terminate).
 *	The number of bytes needed to fully encode the string is returned.
 *
 *	Strvisx encodes exactly len bytes from src into dst.
 *	This is useful for encoding a block of data.
 */
int
strvis(char *dst, const char *src, int flag)
{
	char c;
	char *start;

	for (start = dst; (c = *src); )
		dst = vis(dst, c, flag, *++src);
	*dst = '\0';
	return (dst - start);
}

int
strnvis(char *dst, const char *src, size_t siz, int flag)
{
	char *start, *end;
	char tbuf[5];
	int c, i;

	i = 0;
	for (start = dst, end = start + siz - 1; (c = *src) && dst < end; ) {
		if (isvisible(c)) {
			i = 1;
			*dst++ = c;
			if (c == '\\' && (flag & VIS_NOSLASH) == 0) {
				/* need space for the extra '\\' */
				if (dst < end)
					*dst++ = '\\';
				else {
					dst--;
					i = 2;
					break;
				}
			}
			src++;
		} else {
			i = vis(tbuf, c, flag, *++src) - tbuf;
			if (dst + i <= end) {
				memcpy(dst, tbuf, i);
				dst += i;
			} else {
				src--;
				break;
			}
		}
	}
	if (siz > 0)
		*dst = '\0';
	if (dst + i > end) {
		/* adjust return value for truncation */
		while ((c = *src))
			dst += vis(tbuf, c, flag, *++src) - tbuf;
	}
	return (dst - start);
}

int
strvisx(char *dst, const char *src, size_t len, int flag)
{
	int c;
	char *start;

	for (start = dst; len > 1; len--) {
		c = *src;
		dst = vis(dst, c, flag, *++src);
	}
	if (len)
		dst = vis(dst, *src, flag, '\0');
	*dst = '\0';

	return (dst - start);
}
